### DISCUSSION TITLE
Helix is looking for a plugin system. What is your experience with the Nushell solution?

### DISCUSSION THREAD

**Discussion by getreu** [nushell/nushell#9158] [General]
There is an ongoing [discussion](https://github.com/helix-editor/helix/discussions/3806) about how the future plugin system for Helix should be. Someone [investigated the Nushell solution](https://github.com/helix-editor/helix/discussions/3806#discussioncomment-5315906) and states:

> Nushell plugins use JSON or Protobufs. Not very conserving of resources.

Is this valuation right? Are you satisfied with your solution? What are the pitfalls?

**Comment by ghost** [comment]
I haven't used plugins much, but here is my experience with them:
- with non-rust languages (python, etc): without a library it is annoying to wrap everything into custom datatypes - with a framework its ok, but well..
- with rust: you just use nushell as library, which takes longer to compile and isn't the most stable interface (nushell regularly changes a lot)
- even if you can access all of nushells code some things are missing api interfaces (example: nuon)
- as far as i am aware nushell does lazy-loading (and unloading) in terms of performance -> it shouldnt affect you to much
- it is pretty restricted (you can write commands, but you cant add event handlers as far as i am aware)

but i dont think its fit for a editor:
- nvim plugins are constantly called (CursorMoveEvent, etc) and therefore it would be really ineffective to unload it and keeping a bunch of subprocesses (especially if we mix multiple script languages) around is probably pretty resource intensive
- a stable api interface is basically required (similar to vim.api)
- do you expect every helix user to have python, cargo, etc installed for compiling the plugins (or want to ship binaries)? a restricted language set might be helpful (lua has a pretty good track record with nvim, gmod, etc)

**Comment by fdncred** [comment]
> Is this valuation right? 

I corrected this statement [here](https://github.com/helix-editor/helix/discussions/3806#discussioncomment-5860594).

The protocols we use are not in question. MessagePack is super-fast and JSON is super-portable.

What is in question is a stable ABI to have the functionality we want. JT did quite a bit of research on the variety of plugins including wasm, a couple crates that act as shims to stabilize the ABI, and a variety of other methods. Nothing had exactly what we needed. So, for now, we have our plugin system that has it's pitfalls but we couldn't find anything better that would justify the investment. I really think we're just waiting for a stable ABI. At least that's what I recall from our meetings.
